### PR TITLE
Fix single row prediction performance in a multi-threaded environment

### LINKED ISSUE #6021: Fix single row prediction performance in a multi-threaded environment
## Summary

Single row prediction API takes a write lock on the Booster. This breaks performance when using the same booster in a multi-threaded environment.

## Motivation

The use case is a complex algorithm processing millions of rows and sometimes needing to run a prediction.
The computation is already parallelized at the "millions-of-rows" level.
It needs to use the booster to (sometimes) generate a row and run the prediction on it.
However, currently that performs very poorly due to global contention here:
https://github.com/microsoft/LightGBM/blob/d73c6b530b39a18a3cacaafc4e42550be853c036/src/c_api.cpp#L393

## Description

It seems that this write lock is required due to the fact that the `Predictor` writes inside itself as it predicts (#3771), and that the `Predictor` for single row predictions [is stored inside the `Booster`](https://github.com/microsoft/LightGBM/blob/d73c6b530b39a18a3cacaafc4e42550be853c036/src/c_api.cpp#L376), unlike the `Predictor` for batch prediction functions, which is [instantiated locally by the `Predict` function, allowing locking to be shared](https://github.com/microsoft/LightGBM/blob/d73c6b530b39a18a3cacaafc4e42550be853c036/src/c_api.cpp#L428-L429).

Now the predictor seems to be stored inside the booster in order to be re-usable across calls, to make the single predictions through `LGBM_BoosterPredictForMatSingleRow` reasonably inexpensive: this was implemented before it became recommended to make that a two-step process by using `LGBM_BoosterPredictForMatSingleRowFastInit` instead.
HOWEVER, as of the introduction of `LGBM_BoosterPredictForMatSingleRowFastInit` in #2992, we now have a "SingleRowPredictor" object that is supposed to be instantiated and preserved across calls to predict single rows with `LGBM_BoosterPredictForMatSingleRowFast`: that is `FastConfigHandle`. (Probably would have been better to name it `SingleRowPredictorHandle` as part of the public API but I guess that might be a little late now?)

This means that at the moment, resources necessary for `LGBM_BoosterPredictForMatSingleRowFast` is are split between two places:
 - Some of it is in the booster (the `Predictor`s)
 - Some of it is in the FastConfig (`Config`, [the idea that some single row predictor settings have been set on the booster](https://github.com/microsoft/LightGBM/blob/d73c6b530b39a18a3cacaafc4e42550be853c036/src/c_api.cpp#L2418) (you better hope that nobody else is calling `LGBM_BoosterPredictForMatSingleRowFastInit` with a different config on the same predictor)...)

The idea is: put all the resources that need to be initialized once before running single row predictions behind the `FastConfigHandle` object (aka `SingleRowPredictorHandle`), and take a write lock on that object, instead of the full Booster, when running predictions.

This way, when processing on multiple threads, each thread can make its call to `LGBM_BoosterPredictForMatSingleRowFastInit`, then work without contention:
- Removes the need for taking a global write lock when predicting in single row mode
- Simplifies the booster by removing the `single_row_predictor_` array (should it have been named `single_row_predictors` instead btw? In any case, on `SingleRowPredictorHandle` aka `FastConfigHandle` we only need a single `Predictor` object), and removing the `SetSingleRowPredictor`, `PredictSingleRow` functions
  + Specifically for this, there are two approaches:
     1. Make the `LGBM_BoosterPredictForMatSingleRow` instantiate its `FastConfig` at the beginning of the call. This is basically what it used to do but it didn't hold the predictor, so that would probably decrease performance for users that currently should use `LGBM_BoosterPredictForMatSingleRowFastInit` but use `LGBM_BoosterPredictForMatSingleRow` instead.
     2. Make the `LGBM_BoosterPredictForMatSingleRow` keep using a single row predictor stored in the booster object, but that would be distinct from the one used by `LGBM_BoosterPredictForMatSingleRowFast`. This has the downside of being less clean, because we would need to leave the current `single_row_predictor_`, `SetSingleRowPredictor`, `PredictSingleRow` functions in `Booster`, however that maintains the current performance of `LGBM_BoosterPredictForMatSingleRow` for all users.

cc @AlbertoEAF

### PR DESCRIPTION
**Pull request by Ten0**
Fixes #6021

- Store all resources that are reused across single-row predictions in the dedicated `SingleRowPredictor` (aka `FastConfig`)
- Use that instead of resources in the `Booster` when doing Fast™ single row predictions to avoid having to lock the `Booster` exclusively.

### CODE REVIEW

**Comment by Ten0** [comment]
Workflows run would help :) (not sure what is currently using this and should be tested)

**Comment by Ten0** [comment]
@microsoft-github-policy-service agree

**Comment by jameslamb** [comment]
@shiyu1994 or @guolinke could you please help with a review on this one?

And @AlbertoEAF if you have time?

File: src/c_api.cpp [hunk 1]
```diff
@@ -2053,35 +2113,11 @@ int LGBM_BoosterCalcNumPredict(BoosterHandle handle,
...
-  }
 
-  Booster* const booster;
-  Config config;
-  const int predict_type;
-  const int data_type;
-  const int32_t ncol;
-};
 
 int LGBM_FastConfigFree(FastConfigHandle fastConfig) {
```
Comment (guolinke) [inline src/c_api.cpp:2118]: should we rename this function?

Comment (Ten0) [reply src/c_api.cpp:2118]: Is it not an issue to break backwards compatibility? I renamed internally because it makes more sense with regards to what it does now but opted to not rename public interfaces for that reason.

Comment (guolinke) [reply src/c_api.cpp:2118]: we can add a new function,  and keep LGBM_FastConfigFree for now. And remove the old LGBM_FastConfigFree in the next version.

Comment (Ten0) [reply src/c_api.cpp:2118]: So same thing for the two fast config init functions as well I suppose?

Comment (AlbertoEAF) [reply src/c_api.cpp:2118]: Hello, sorry I didn't have time to review this carefully, but if we can keep the public API for now the better 👍 . Some external libraries were using this part of the API already.

**Comment by AlbertoEAF** [comment]
@Ten0 thanks for contributing! Since we're adding capability for it, can you add a test to our C++ test suite showing multiple predictors using the fast API and sharing the booster?

**Comment by Ten0** [comment]
So I've finished my benchmarks and it turns out:
- I can see the expected performance improvements of x N_CPUs performance
- [The sub-optimal locking](https://github.com/microsoft/LightGBM/pull/2760#issuecomment-1683075872) has no impact
- I was afraid that memory barriers due to the read lock might be noticeable, so I thought a bit unfortunate that we would have to take the shared read lock for every prediction. I was originally considering holding the shared lock for the entire lifetime of the SingleRowPredictor, but as it turned out some implementations using the C library (notably R) would rely on being able to update the Booster after constructing a SingleRowPredictor, so that would cause a deadlock. (https://github.com/microsoft/LightGBM/pull/6024/commits/f41755b42b37f16c5dfc43c55cdc78ef41c463db). As it turns out, this shows no impact in my benchmarks, so I'm going to be reasonably happy with leaving it as is (that is, with taking a shared lock for every single row prediction, as is currently implemented).
  image

File: src/c_api.cpp (hunk 1 above)
Comment (Ten0) [reply src/c_api.cpp:2118]: So I'm not sure I understand whether we should create the new functions *and* keep the old ones, or completely leave the public API as is.

Comment (guolinke) [reply src/c_api.cpp:2118]: sorry for the late response. I think at least we should add a TODO tag here, and update this API name in next (or next next) official released version.

**Comment by stonebrakert6** [comment]
Thanks a lot @Ten0 for this commit. I am a new user of lightGBM(need to write this so that you don't confuse me with a veteran).
So, IIRC, the usage now would be something similar to this

1. `BoosterHandle` is created once and is then shared among all threads.
2. Each thread has its own copy of `FastConfigHandle` and then calls `LGBM_BoosterPredictForMatSingleRowFastInit` to initialize its private(non-shared) `FastConfigHandle` using the `BoosterHandle` created in step 1.
3. Each thread can now call `LGBM_BoosterPredictForMatSingleRowFast`(passing in its `FastConfigHandle`) to make predictions.

Is this understanding correct?

**Comment by Ten0** [comment]
> Each thread has its own copy of `FastConfigHandle`

Of `BoosterHandle` but I think that's what you meant. Otherwise yes that's exactly it.

**Comment by stonebrakert6** [comment]
> Of BoosterHandle but I think that's what you meant. 

Sorry, but I didn't mean that. I really meant `FastConfigHandle`. Quoting you on #6021 

> This way, when processing on multiple threads, each thread can make its call to LGBM_BoosterPredictForMatSingleRowFastInit, then work without contention:


Here is the pseudo-code, would need to fill `...` to make it compile. Also, it has memory/resource leakage.
```cpp
#include <iostream>
#include <thread>
#include <vector>

#include "LightGBM/c_api.h"

// # of features
const int kFeatures = 13;
// # of threads which call predict
const int nworkers = 10;
// Shared booster
BoosterHandle booster_handle;

// Every thread exectues this function
void predict(ssize_t beg, ssize_t end) {
  // Each thread has its own FastConfigHandle
  FastConfigHandle config;
  int rc = LGBM_BoosterPredictForMatSingleRowFastInit(
      booster_handle, C_API_PREDICT_NORMAL, 0, 0, C_API_DTYPE_FLOAT64,
      kFeatures, "", &config);
  if (rc != 0) {
    abort();
  }
  for (ssize_t i = beg; i < end; i++) {
    int64_t len = 0;
    rc = LGBM_BoosterPredictForMatSingleRowFast(config, ...);
  }
}

int main(int argc, char* argv[]) {
  int num_iterations;
  int rc = LGBM_BoosterCreateFromModelfile(argv[1], &num_iterations,
                                           &booster_handle);
  if (rc != 0) {
    std::cout << "LGBM_BoosterCreateFromModelfile() returned " << rc << '\n';
    return 1;
  }
  std::vector<std::thread> workers(nworkers);
  for (ssize_t i = 0; i < nworkers; i++) {
    if (i != nworkers - 1) {
      workers[i] = std::thread(predict, ...);
    }
  }
  for (std::thread& t : workers) {
    t.join();
  }
  return 0;
}

```
If each thread creates its own `Booster` then memory usage for a single model would be x N, where N is the number of threads. So I want to share the booster with all threads and during prediction, I don't want to acquire the write lock, (that's why I am using your fork).

Let me know if my understanding of API usage is correct or not. I want to use your fork(because I need scalable predict/inference performance)

**Comment by Ten0** [comment]
> Let me know if my understanding of API usage is correct or not

Your understanding is correct, it seems to be just a matter of definitions that caused our misunderstanding. (`Copy`, handle, globals...)

> I want to use your fork(because I need scalable predict/inference performance)

Happy to help :)
If you happen to feel like helping to get this merged to not have to deal with the complications of using a fork I'd happily accept contributions to the PR with regards to [the "please showcase usage in the C++ test suite" review comment](https://github.com/microsoft/LightGBM/pull/6024#issuecomment-1712823369) - I'm not sure exactly when I'll have time to finalize this.

**Comment by stonebrakert6** [comment]
Okay, 

I think with the above-described usage, the API `LGBM_BoosterPredictForMatSingleRowFastInit` is not thread-safe. 
Here is the chain of function/method calls
`LGBM_BoosterPredictForMatSingleRowFastInit()` -> `Booster::InitSingleRowPredictor()` -> `SingleRowPredictor::SingleRowPredictor()` ->  `SingleRowPredictorInner::SingleRowPredictorInner()` -> `Predictor::Predictor()` -> `GBDT::InitPredict()` which writes concurrently i.e data race(or overwrites if called serially) to variables(atleast) `num_iteration_for_pred_` and `start_iteration_for_pred_` of Boosting object(concretely `GBDT`) `src/boosting/gbdt.h:422`

Please ignore this as noise if you think this is a false claim.
NOTE: I could be completely wrong as I just started evaluating this library for a project at my workplace and my focus is only on scalable inference/predict performance, given an ML model trained by our ML team(who have selected this library). I have very shallow knowledge of ML concepts and my focus is only to ensure that inference for trained model scales for the application that would use this model only for inference. So I was pretty surprised to see the write lock for prediction considering that we aren't really "learning" or writing to the model, but as you very well described in your explanation at #6021 that 
> Predictor writes inside itself as it predicts

> I'd happily accept contributions to the PR with regards to https://github.com/microsoft/LightGBM/pull/6024#issuecomment-1712823369

I'll be happy to contribute once I get some feet wet with the codebase.

Thanks a lot for your patience.

**Comment by stonebrakert6** [comment]
Okay, 

I think with the above-described usage, the API `LGBM_BoosterPredictForMatSingleRowFastInit` is not thread-safe. 
Here is the chain of function/method calls
`LGBM_BoosterPredictForMatSingleRowFastInit()` -> `Booster::InitSingleRowPredictor()` -> `SingleRowPredictor::SingleRowPredictor()` ->  `SingleRowPredictorInner::SingleRowPredictorInner()` -> `Predictor::Predictor()` -> `GBDT::InitPredict()` which writes concurrently i.e data race to variables(atleast) `num_iteration_for_pred_` and `start_iteration_for_pred_` of Boosting object(concretely `GBDT`) `src/boosting/gbdt.h:422`

Feel free to ignore(or delete this message) this if you think this is not problematic or false and causing noise in the PR by being at the wrong place(should be at fork?)

NOTE: I just started evaluating this library for a project at my workplace. Our ML team has selected this library and they are training the model offline while our application is only concerned with inference/prediction on the trained model.  My knowledge of ML is pretty shallow but my job is to ensure that it scales for prediction/inference since our application is extremely latency-sensitive. So I was pretty shocked to see the write lock(why do I need a write lock to make inference?), but you explained it very well at #6021 that
> Predictor writes inside itself as it predicts 


>  I'd happily accept contributions to the PR with regards to the "please showcase usage in the C++ test suite" review commen

I'll be happy to contribute once I get some familiarity with the code base.
Thanks for your patience.

**Comment by Ten0** [comment]
> I think with the above-described usage, the API `LGBM_BoosterPredictForMatSingleRowFastInit` is not thread-safe. Feel free to ignore(or delete this message) this if you think this is not problematic or false and causing noise in the PR by being at the wrong place(should be at fork?)

I did not particularly check anything beyond `Predictor::Predictor` because anything beyond that point can already be called concurrently in the current version of the library, through the following call chain:
[`LGBM_BoosterPredictForMat()`](https://github.com/microsoft/LightGBM/blob/0f7983b6c3443154441cecaa342462d4567760b7/src/c_api.cpp#L2372) -> [`Booster::Predict()`](https://github.com/microsoft/LightGBM/blob/0f7983b6c3443154441cecaa342462d4567760b7/src/c_api.cpp#L428-L429) -> [`Booster::CreatePredictor()`](https://github.com/microsoft/LightGBM/blob/0f7983b6c3443154441cecaa342462d4567760b7/src/c_api.cpp#L420-L421) -> [`Predictor::Predictor()`](https://github.com/microsoft/LightGBM/blob/b0137debe6e9cc92b65ec71b0fe8a56ea213c143/src/application/predictor.hpp#L61)

So it seems that if the issue you describe is indeed an issue (which it seems to be AFAICT), then it pre-dates this PR. In that case, it seems to be beyond my scope, and we should probably open a separate issue for this. (Bug was probably introduced as soon as #2760)

> my job is to ensure that it scales for prediction/inference since our application is extremely latency-sensitive

I am actually in a pretty similar situation 🤗

File: src/c_api.cpp (hunk 1 above)
Comment (Ten0) [reply src/c_api.cpp:2124]: 6d949a0b910b1a6b633b228cc7947c2332d2bb95

**Comment by Ten0** [comment]
TLDR; All comments resolved.

> can you add a test to our C++ test suite showing multiple predictors using the fast API and sharing the booster?

Ah well that took a while to implement because a similar test didn't already exist (the C api was only tested through the python wrapper and the other implementations themselves) but it's finally done!
1fcbc3fd7fed762e1b753bdccbabebf644d63667

I've also added a workaround for #6142. The state of this issue is that, with the workaround, users would not hit issues any more with this API than they would with the current existing APIs around both single row and matrix prediction.

AFAICT this is ready for merge.

**Comment by Ten0** [comment]
@guolinke Static analysis complains that the TODO tag is not assigned. Who should I assign it to?

**Comment by jameslamb** [comment]
> TODO

We prefer not using `TODO` comments at all, and instead using the Issues backlog to track future work. You won't find many in this repo:

* https://github.com/search?q=repo%3Amicrosoft%2FLightGBM%20TODO&type=code
* https://github.com/search?q=repo%3Amicrosoft%2FLightGBM+FIXME&type=code

Could you please put the content of that comment into a new issue at https://github.com/microsoft/LightGBM/issues with as much detail as necessary for someone else to pick it up (and linking to this PR)?

**Review by guolinke — APPROVED** [review]
Thank you!

**Comment by jameslamb** [comment]
Sorry, I'd missed that @guolinke approved this back in January!

I've just updated it to latest `master`. Will merge this today assuming that CI passes.

**Review by jameslamb — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks very much!
